karapace/protobuf/serialization.py

```diff
@@ -0,0 +1,316 @@
+    public_imports: list[TypeName] = []
+    for index, dep in enumerate(proto.dependency):
+        if index in proto.public_dependency:
+            public_imports.append(TypeName(dep))
+        else:
+            imports.append(TypeName(dep))
+    types: list[TypeElement] = []
+    for enumtype in proto.enum_type:
+        types.append(_deserialize_enum(enumtype))
+    for msgtype in proto.message_type:
+        types.append(_deserialize_msg(msgtype))
+    options: list[OptionElement] = _deserialize_options(proto.options)
+
+    return ProtoFileElement(
+        DEFAULT_LOCATION,
+        package_name=PackageName(proto.package),
+        syntax=Syntax(proto.syntax),
+        imports=imports,
+        public_imports=public_imports,
+        types=types,
+        options=options,
+    )
+
+
+_LABEL_MAP = MappingProxyType(
+    {
+        Field.Label.OPTIONAL: google.protobuf.descriptor_pb2.FieldDescriptorProto.LABEL_OPTIONAL,
+        Field.Label.REQUIRED: google.protobuf.descriptor_pb2.FieldDescriptorProto.LABEL_REQUIRED,
+        Field.Label.REPEATED: google.protobuf.descriptor_pb2.FieldDescriptorProto.LABEL_REPEATED,
+    }
+)
+
+
+def _serialize_field_label(label: Field.Label) -> google.protobuf.descriptor_pb2.FieldDescriptorProto.Label.ValueType:
+    if label not in _LABEL_MAP:
+        raise NotImplementedError(f"Unsupported field label {label}")
+    return _LABEL_MAP[label]
+
+
+def _serialize_field(field: FieldElement) -> google.protobuf.descriptor_pb2.FieldDescriptorProto:
+    d = google.protobuf.descriptor_pb2.FieldDescriptorProto()
+    if field.label is not None:
+        d.label = _serialize_field_label(field.label)
+        if field.label == Field.Label.OPTIONAL:
+            d.proto3_optional = True
+    else:
+        d.label = _serialize_field_label(Field.Label.OPTIONAL)
+    if _REVERSE_TYPE_MAP.get(field.element_type) is not None:
+        d.type = _REVERSE_TYPE_MAP[field.element_type]
+    else:
+        d.type_name = field.element_type
+    if field.name is not None:
+        d.name = field.name
+    if field.tag is not None:
+        d.number = field.tag
+    return d
+
+
+def _serialize_enumtype(e: EnumElement) -> google.protobuf.descriptor_pb2.EnumDescriptorProto:
+    result = google.protobuf.descriptor_pb2.EnumDescriptorProto()
+    result.name = e.name
+    for c in e.constants:
+        c2 = google.protobuf.descriptor_pb2.EnumValueDescriptorProto()
+        c2.name = c.name
+        c2.number = c.tag
+        result.value.append(c2)
+    return result
+
+
+def _serialize_type(t: TypeElement) -> google.protobuf.descriptor_pb2.DescriptorProto:
+    d = google.protobuf.descriptor_pb2.DescriptorProto()
+    d.name = t.name
+    for nt in t.nested_types:
+        if isinstance(nt, MessageElement):
+            d.nested_type.append(_serialize_type(nt))
+        elif isinstance(nt, EnumElement):
+            d.enum_type.append(_serialize_enumtype(nt))
+    if isinstance(t, MessageElement):
```

Comment:
What would the else be here?

Reply:
Added else branch throwing an exception

Reply:
Ok, understood your point wrongly first. Now fixed by renaming `_serialize_type` -> `_serialize_msgtype` and only takes `MessageElement`